Haptics/app/src/main/java/com/example/android/haptics/samples/ui/wobble/WobbleRoute.kt

```diff
@@ -0,0 +1,235 @@
+                            dragDistance = maxDragDistance
+                        }
+                    }
+                )
+            ) {
+                Box(Modifier
+                        .align(Alignment.TopCenter)
+                        .offset(y = 150.dp)) {
+                    androidx.compose.animation.AnimatedVisibility(
+                        enter = fadeIn(),
+                        exit = fadeOut(),
+                        visible = dragDistance === DRAG_DISTANCE_START && !isWobbling
+                    ) {
+                        WobbleInstructions()
+                    }
+                }
+                WobbleShape(WOBBLE_SHAPE_HEIGHT,
+                            elasticTopPercent = dragDistanceAnimated / maxDragDistance)
+            }
+        }
+    }
+
+/**
+ * Wobble shape is composed of an ElasticTop shape on top of a rectangle.
+ */
+@Composable
+private fun WobbleShape(wobbleShapeHeight: Dp, elasticTopPercent: Float) {
+    Column(Modifier.fillMaxSize(), verticalArrangement = Arrangement.Bottom) {
+        val halfOfTotalWobbleShapeHeight = wobbleShapeHeight / 2
+        Box(
+            modifier = Modifier
+                .fillMaxWidth()
+                .height(halfOfTotalWobbleShapeHeight)
+                .clip(ElasticTopShape(elasticTopPercent))
+                .background(MaterialTheme.colors.primaryVariant)
+        )
+        Box(
+            modifier = Modifier
+                .fillMaxWidth()
+                .height(halfOfTotalWobbleShapeHeight)
+                .clip(RectangleShape)
+                .background(MaterialTheme.colors.primaryVariant)
+        )
+    }
+}
+
+/**
+ * Instructions for the user to interact with the WobbleShape.
+ */
+@Composable
+private fun WobbleInstructions() {
+    Column(horizontalAlignment = Alignment.CenterHorizontally) {
+        Text(stringResource(R.string.wobble_drag_and_release))
+        Spacer(modifier = Modifier.height(16.dp))
+        Icon(Icons.Rounded.ArrowDownward, null, tint = MaterialTheme.colors.secondaryText)
+    }
+}
+
+
+private fun randomIntensity(baseIntensity: Float): Float {
+    return (baseIntensity + nextOffset()).coerceIn(0f, 1f)
+}
+
+private fun nextOffset(): Float {
+    return RANDOM.nextFloat() * SPIN_RANDOM_WINDOW - SPIN_RANDOM_WINDOW / 2f
+}
+
+private fun vibrate(vibrator: Vibrator, intensity: Float) {
+    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S) {
+        // Spin is only supported in Android S.
+        return
+    }
+    vibrator.vibrate(
+        VibrationEffect.startComposition()
+            .addPrimitive(
+                VibrationEffect.Composition.PRIMITIVE_SPIN,
+                randomIntensity(intensity)
+            )
+            .addPrimitive(
+                VibrationEffect.Composition.PRIMITIVE_SPIN,
+                randomIntensity(intensity)
+            )
+            .addPrimitive(
+                VibrationEffect.Composition.PRIMITIVE_SPIN,
```

Comment:
I think that with a third spin then the duration between vibrate calls need to be updated. Maybe we can load the SPIN primitive duration and use it as well? Since we expect S+ then we don't even need the default value of 260. I'd probably add a padding of 20ms between vibrate calls, so delay = 3 * spinDuration + 20

WDYT?

Reply:
Done!